chain/network/src/peer_manager.rs

```diff
@@ -105,6 +140,167 @@ struct ActivePeer {
+    #[cfg(feature = "adversarial")]
+    GetIbfSet(PeerId),
+    PartialSync {
+        peer_id: PeerId,
+        seed: u64,
+        ibf_level: ValidIBFLevel,
+        ibf: Vec<IbfBox>,
+    },
+    RequestMissingEdges {
+        peer_id: PeerId,
+        requested_edges: Vec<u64>,
+    },
+}
+
+#[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+impl Message for IbfSetConstructorMessages {
+    type Result = IbfSetConstructorMessagesResponse;
+}
+
+#[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+#[derive(MessageResponse, Debug)]
+pub enum IbfSetConstructorMessagesResponse {
+    AddIbfSetResponse {
+        seed: u64,
+        ibf_vec: Vec<IbfBox>,
+    },
+    PartialSyncResponse {
+        edges_for_peer: Vec<SimpleEdge>,
+        unknown_edge_hashes: Vec<u64>,
+        unknown_edges_count: u64,
+        seed: u64,
+        ibf_set: Arc<Mutex<IbfSet<SimpleEdge>>>,
+    },
+    RequestMissingEdgesResponse {
+        edges_for_peer: Vec<SimpleEdge>,
+    },
+    #[cfg(feature = "adversarial")]
+    GetIbfSetResponse { seed: Option<u64> },
+    Empty(),
+}
+
+#[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+impl IbfSetConstructor {
+    pub fn exchange_routing_tables_using_ibf(
+        &self,
+        peer_id: &PeerId,
+        ibf_set: Arc<Mutex<IbfSet<SimpleEdge>>>,
+        ibf_level: ValidIBFLevel,
+        ibf_vec: &[IbfBox],
+        seed: u64,
+    ) -> (Vec<SimpleEdge>, Vec<u64>, u64) {
+        let ibf = ibf_set.lock().unwrap().get_ibf(ibf_level);
+
+        let mut new_ibf = Ibf::from_vec(ibf_vec.clone(), seed ^ (ibf_level.0 as u64));
+
+        if !new_ibf.merge(&ibf.data, seed ^ (ibf_level.0 as u64)) {
+            error!(target: "network", "exchange routing tables failed with peer {}", peer_id);
+            return (Default::default(), Default::default(), 0);
+        }
+
+        let (edge_hashes, unknown_edges_count) = new_ibf.try_recover();
+
+        error!("PIOTR len {:?} {:?}", edge_hashes.len(), unknown_edges_count);
+        let (known, unknown_edges) = self.split_edges_for_peer(&peer_id, &edge_hashes);
+
+        (known, unknown_edges, unknown_edges_count)
+    }
+}
+#[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+impl Handler<IbfSetConstructorMessages> for IbfSetConstructor {
+    type Result = IbfSetConstructorMessagesResponse;
+
+    #[perf]
+    fn handle(&mut self, msg: IbfSetConstructorMessages, _ctx: &mut Self::Context) -> Self::Result {
+        match msg {
+            IbfSetConstructorMessages::AddEdges(edges) => {
+                for e in edges.iter() {
+                    let se = e.to_simple_edge();
+                    self.peer_ibf_set.add_edge(&se);
+                }
+            }
+            #[cfg(feature = "adversarial")]
+            IbfSetConstructorMessages::RemoveEdges(edges) => {
+                for edge in edges.iter() {
+                    self.peer_ibf_set.remove_edge(edge);
+                }
+            }
```

Comment:
Is it intended that edges can only be removed when `adversarial` flag is enabled?

Reply:
@bowenwang1996 Yes, you are right. We should be calling RemoveEdges after doing edge pruning as well.

Reply:
Thanks, fixed.

Reply:
Why is this not caught by our test? Could you incorporate this into the test you wrote?

Reply:
We were not calling that code path in the test. I changed tests to use it.